Remove redundant temp folder and unused imports

MVrachev opened the pull request:
**Description of the changes being introduced by the pull request**:

In many of the tests classes, we are creating two
temporary directories: one with the name "temp_<random_string>"
and inside it, we generate a new directory for each test with the name
`Test<Class_Name>_<random_string>`.

I think we don't need the "temp_<random_string>" directory.
The only benefit I can think of is that it could contain multiple
temp folders from failing tests from a particular test run.
But even then, this is not a big bonus because the name
`temp_<random_string>` is not really descriptive from which test
run this directory was created.

PS: Thanks to Jussi Kukkonen who noticed we are using two temp
folders per class in our tests.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

**Please verify and check that the pull request fulfills the following
requirements**:

- [ ] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

Commit:
Remove redundant temp folder and unused imports

In many of the tests classes, we are creating two
temporary directories: one with the name "temp_<random_string>"
and inside it, we generate a new directory for each test with the name
"Test<Class_Name>_<random_string>".

I think we don't need the "temp_<random_string>" directory.
The only benefit I can think of is that it could contain multiple
temp folders from failing tests from a particular test run.
But even then, this is not a big bonus because the name
"temp_<random_string>" is not really descriptive from which test
run this directory was created.

Also, fixed the way we import "unittest_toolbox" and replace
"unittest_toolbox.Modified_TestCase.function(self)"
with "super().function()".

PS: Thanks to Jussi Kukkonen who noticed we are using two temp
folders per class in our tests.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

joshuagl commented on tests/test_repository_lib.py:
```diff
@@ -86,24 +81,25 @@ def tearDownClass(cls):
 
   def setUp(self):
+    # We are inheriting from custom class.
+    unittest_toolbox.Modified_TestCase.setUp(self)
```
This pattern looks odd, isn't the normal way to do this with `super().setUp(self)` ?

joshuagl commented on tests/test_arbitrary_package_attack.py:
```diff
@@ -102,8 +92,10 @@ def setUp(self):
     # any modifications made to repository files are restricted to the copies.
     # The 'repository_data' directory is expected to exist in 'tuf/tests/'.
     original_repository_files = os.path.join(os.getcwd(), 'repository_data')
+    # Create a temporary directory to store the repository, metadata,
+    # and target files.
```
I don't think we need to include this comment here or throughout the tests you've edited.

joshuagl commented on tests/test_repository_lib.py:
```diff
@@ -86,24 +81,25 @@ def tearDownClass(cls):
     tuf.roledb.clear_roledb(clear_all=True)
     tuf.keydb.clear_keydb(clear_all=True)
-
+    # Modified_TestCase.tearDown() automatically deletes temporary files and
+    # directories that may have been created during each test case.
```
Feels like an unnecessary comment throughout the PR

joshuagl commented on tests/test_repository_tool.py:
```diff
@@ -42,6 +42,7 @@
 import tuf.formats
 import tuf.roledb
 import tuf.keydb
+import tuf.unittest_toolbox as unittest_toolbox
```
`from tuf import unittest_toolbox`

joshuagl commented on tests/test_repository_lib.py:
```diff
@@ -42,11 +41,12 @@
 import tuf.keydb
 import tuf.settings
 
+import tuf.unittest_toolbox as unittest_toolbox
```
`from tuf import unittest_toolbox` should suffice

joshuagl commented on tests/test_root_versioning_integration.py:
```diff
@@ -38,6 +38,7 @@
 import tuf.exceptions
 import tuf.roledb
 import tuf.keydb
+import tuf.unittest_toolbox as unittest_toolbox
```
`from tuf import unittest_toolbox`?

joshuagl commented on tests/test_repository_lib.py:
```diff
@@ -86,24 +81,25 @@ def tearDownClass(cls):
-
 
   def setUp(self):
+    # We are inheriting from custom class.
```
I don't think we need the comment to justify chaining up to the parent implementation of the method here, or throughout the tests you have edited.

joshuagl reviewed (commented):
Thank you for taking the time to clean this up. I have a few minor comments, mostly on not adding semi-redundant comments to the code.

MVrachev commented:
Addressed all of @joshuagl comments and:
1. Replaced `import tuf.unittest_toolbox as unittest_toolbox` with `from tuf import unittest_toolbox` in all of our test files.
2. Removed the comments I have added such as `We are inheriting from custom class.` from all of the tests.
3. Replaced all `unittest_toolbox.Modified_TestCase.function(self)` calls with `super().function()` in all of our tests.

MVrachev commented:
It seems like because we are supporting `Python 2` we have a problem with 
`super().function()` which I used because in Python2  it is `super(self).function()`
Probably I should replace it with `unittest_toolbox.Modified_TestCase.function(self)` the way it was?

joshuagl commented:
Ah, yes. The syntax changed for `super()` between Python 2 and Python 3. Let's leave it as is for now and we can fix this later when we've purged Python 2 from our codebase.

joshuagl commented:
Would be good to get this cleanup landed. @MVrachev could you revert back to `unittest_toolbox.Modified_TestCase.function(self)` when you have some cycles? Thank you.

jku commented:
I'm a bit worried about
* using os.getcwd() for the location where temp dirs are created
* the fact that sometimes temp dirs are not cleaned up (is this only when test run is cancelled? not sure)
* after this PR probably creating more temp dirs in current dir (one per instance instead of one per class I guess?)

So I wonder if this should only be done if we also start creating the temp dirs somewhere more reasonable?

Does anyone know when the temp dirs are not being removed?

sechkova commented:
> Does anyone know when the temp dirs are not being removed?

Besides the expected case when `tearDown` is not called due to some error, I've noticed temp dirs not being removed on Windows with "being used by another process" error. Unfortunately, I can't say which is the process and I haven't observed this consistently in other environments.

sechkova commented:
> So I wonder if this should only be done if we also start creating the temp dirs somewhere more reasonable?

I would vote for that if possible. Polluting the tests source code dir on a failure is not particularly harmful but ... unpleasant

MVrachev commented:
> I'm a bit worried about
> 
> * using os.getcwd() for the location where temp dirs are created
>
> * the fact that sometimes temp dirs are not cleaned up (is this only when test run is cancelled? not sure)


> * after this PR probably creating more temp dirs in current dir (one per instance instead of one per class I guess?)

Yes, the side effect is that it will create one temp dir per class instance.

> So I wonder if this should only be done if we also start creating the temp dirs somewhere more reasonable?

As I understand it, the idea is to have those temp files available for you on test failure for analysis.
Probably it will be better to choose a deterministic path where they can be found, but where should that be?
Maybe `~/tmp?`


> Does anyone know when the temp dirs are not being removed?

The [doc](https://docs.python.org/3/library/unittest.html#unittest.TestCase.tearDown) says:
`tearDown`:
```
Method called immediately after the test method has been called and the result recorded. 
This is called even if the test method raised an exception, so the implementation 
in subclasses may need to be particularly careful about checking the internal state.
...
This method will only be called if the setUp() succeeds, regardless of the outcome 
of the test method. 
```

Again from the [doc](https://docs.python.org/3/library/unittest.html#setupclass-and-teardownclass) for tearDownClass:
```
If an exception is raised during a setUpClass then the tests in the class are not run and 
the tearDownClass is not run. 
```

Also again in the [doc](https://docs.python.org/3/library/unittest.html#skipping-tests-and-expected-failures)
```
Skipped tests will not have setUp() or tearDown() run around them. 
Skipped classes will not have setUpClass() or tearDownClass() run.
```

joshuagl commented:
> As I understand it, the idea is to have those temp files available for you on test failure for analysis.
> Probably it will be better to choose a deterministic path where they can be found, but where should that be?
> Maybe `~/tmp?`

Python has a function to get an OS appropriate, administrator configurable, temporary directory [`tempfile.gettempdir()`](https://docs.python.org/3/library/tempfile.html?tempfile.mkdtemp#tempfile.gettempdir). As well as providing functions to create directories within that host appropriate location: [`tempfile.mkdtemp()`](https://docs.python.org/3/library/tempfile.html?tempfile.mkdtemp#tempfile.mkdtemp), and [`tempfile.TemporaryFile()`](https://docs.python.org/3/library/tempfile.html?tempfile.mkdtemp#tempfile.TemporaryFile).

sechkova commented:
> > Does anyone know when the temp dirs are not being removed?
> 
> Besides the expected case when `tearDown` is not called due to some error, I've noticed temp dirs not being removed on Windows with "being used by another process" error. Unfortunately, I can't say which is the process and I haven't observed this consistently in other environments.

Sorry for giving you a bit random hints and not direct answers but in the Windows case temporary files created by `unittest_toolbox.Modified_TestCase.make_temp_data_file` cannot be cleaned due to 
`PermissionError(13, 'This process cannot access the file because it is being used by another process')`. 

But this is a Windows VM so it can be something specific to this case ...

MVrachev commented:
Blocked, until we remove Python2 from our tests as Joshua has mentioned.
https://github.com/theupdateframework/tuf/pull/1224#issuecomment-738333840

MVrachev commented:
After the removal of python2 test cases in https://github.com/theupdateframework/tuf/commit/13b085712f2b4b83fc2e7014c34016d55061ab7f I rebased and updated this pr.

MVrachev commented:
Honestly, I am wondering about this change.
On one hand, on a successful test run, we had generated multiple additional folders for each test class,
on another hand, on an error probably we will end up with many folders for each test class that failed, right?

jku commented:
I might have broken this PR quite badly with a recent bug fix (moved chained up teardown calls to the correct place in #1346 to fix issues found in another PR). Sorry about that -- I did not remember this one.

I think that PR also makes this one a bit less needed: at least in the tests that I touched I made sure that we only try to remove the "topmost" directory (because that will remove every other directory inside it). In some cases there might still be unneeded temp directories.

Let's discuss when you are back.

MVrachev commented:
As I said before, I don't feel so confident about this change anymore.
I will just close it for now and if somebody complains about it in the future, I will reopen or push another one.
